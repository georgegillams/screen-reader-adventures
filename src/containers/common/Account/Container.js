import React from 'react';
import PropTypes from 'prop-types';
import PageTitle from 'components/common/PageTitle';
import { LoadingCover } from 'gg-components/LoadingCover';
import { DebugObject } from 'components/common/DebugObject';
import { Paragraph } from 'gg-components/Paragraph';
import Button from 'components/common/Button';
import { setPostLoginRedirect } from 'client-utils/common/storageHelpers';

import Skeleton from './Skeleton';

import { LoggedInOnly } from 'components/common/Walls';
import STYLES from './account.scss';
import { cssModules } from 'gg-components/helpers/cssModules';

const getClassName = cssModules(STYLES);

const Account = props => {
  const {
    logout,
    requestVerificationEmail,

    accountState,
    authenticatorState,

    className,
  } = props;
  const { user } = authenticatorState;

  const outerClassNames = [];

  if (className) {
    outerClassNames.push(className);
  }

  const showEmail = user && user.email;
  const showUname = user && user.uname;
  const showUserDetails = showEmail || showUname;

  const page = (
    <div className={outerClassNames.join(' ')}>
      <LoggedInOnly
        user={authenticatorState.user}
        setLoginRedirect={() => {
          setPostLoginRedirect('account');
        }}>
        <PageTitle name="Account">
          {showUserDetails && (
            <Paragraph className={getClassName('account__details')}>
              {showEmail && <div>{`Email: ${user.email}`}</div>}
              {showUname && <div>{`Display name: ${user.uname}`}</div>}
            </Paragraph>
          )}
          <div className={getClassName('account__control-panel')}>
            {user && !user.emailVerified && (
              <>
                <Button
                  className={getClassName('account__button')}
                  loading={accountState && accountState.requestingVerificationEmail}
                  onClick={requestVerificationEmail}>
                  Get a new verification email
                </Button>
              </>
            )}
            {user && user.admin && (
              <>
                <Button className={getClassName('account__button')} href="/admin">
                  Admin
                </Button>
                <Button className={getClassName('account__button')} href="/status">
                  Site status
                </Button>
              </>
            )}
            <Button destructive loading={accountState && accountState.loggingOut} onClick={logout}>
              Logout
            </Button>
          </div>
        </PageTitle>
      </LoggedInOnly>
    </div>
  );

  return (
    <>
      <LoadingCover loadingSkeleton={Skeleton} loading={authenticatorState.user === undefined} error={false}>
        {page}
      </LoadingCover>
      <DebugObject
        debugTitle="Account"
        debugObject={{
          logout,
          authenticatorState,
          accountState,
        }}
      />
    </>
  );
};

Account.propTypes = {
  logout: PropTypes.func.isRequired,
  requestVerificationEmail: PropTypes.func.isRequired,
  accountState: PropTypes.shape({
    loggingOut: PropTypes.bool,
    logOutError: PropTypes.object,
    logOutResult: PropTypes.object,
    requestingVerificationEmail: PropTypes.bool,
    requestVerificationEmailError: PropTypes.object,
    requestVerificationEmailResult: PropTypes.object,
  }).isRequired,
  authenticatorState: PropTypes.shape({
    user: PropTypes.object,
  }).isRequired,
  className: PropTypes.string,
};

Account.defaultProps = {
  authenticatorState: null,
  logoutState: null,
  className: null,
};

export default Account;
