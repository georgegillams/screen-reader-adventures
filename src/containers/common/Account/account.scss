@import '~utils/tokens';

.account {
  &__details {
    display: inline-block;
    width: 100%;
    margin-bottom: $spacing-base;
  }

  &__button {
    margin-bottom: $spacing-base;
  }

  &__control-panel {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
  }
}
