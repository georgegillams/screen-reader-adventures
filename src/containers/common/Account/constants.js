export const KEY = 'account';
export const LOGOUT = `app/${KEY}/LOGOUT`;
export const REQUEST_VERIFICATION_EMAIL = `app/${KEY}/REQUEST_VERIFICATION_EMAIL`;
