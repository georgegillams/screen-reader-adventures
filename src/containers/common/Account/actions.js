import { createRoutine } from 'redux-saga-routines';

import { LOGOUT, REQUEST_VERIFICATION_EMAIL } from './constants';

export const logout = createRoutine(LOGOUT);
export const requestVerificationEmail = createRoutine(REQUEST_VERIFICATION_EMAIL);
