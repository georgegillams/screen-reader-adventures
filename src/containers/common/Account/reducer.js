import produce from 'immer';

import { logout, requestVerificationEmail } from './actions';

export const initialState = {
  loggingOut: false,
  logOutError: null,
  logOutResult: null,
  requestingVerificationEmail: false,
  requestVerificationEmailError: null,
  requestVerificationEmailResult: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case logout.REQUEST:
        draft.loggingOut = true;
        draft.logOutError = null;
        break;

      case logout.SUCCESS:
        draft.loggingOut = false;
        draft.logOutResult = payload;
        break;

      case logout.FAILURE:
        draft.loggingOut = false;
        draft.logOutError = payload;
        break;

      case requestVerificationEmail.REQUEST:
        draft.requestingVerificationEmail = true;
        draft.requestVerificationEmailError = null;
        break;

      case requestVerificationEmail.SUCCESS:
        draft.requestingVerificationEmail = false;
        draft.requestVerificationEmailResult = payload;
        break;

      case requestVerificationEmail.FAILURE:
        draft.requestingVerificationEmail = false;
        draft.requestVerificationEmailError = payload;
        break;
    }
  });

export default reducer;
