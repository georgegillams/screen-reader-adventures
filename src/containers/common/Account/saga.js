import { takeLatest, put, call } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { logout, requestVerificationEmail } from './actions';
import { setUser } from 'containers/common/Authenticator/actions';

export function* doLogout() {
  const requestURL = apiStructure.logout.fullPath;

  try {
    yield put(logout.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
    });

    if (result.error) {
      yield put(logout.failure(result));
    } else {
      yield put(logout.success(result));
      yield put(setUser.trigger(null));
    }
  } catch (err) {
    yield put(logout.failure(err));
  } finally {
    yield put(logout.fulfill());
  }
}

export function* doRequestVerificationEmail() {
  const requestURL = apiStructure.requestVerificationEmail.fullPath;

  try {
    yield put(requestVerificationEmail.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
    });

    if (result.error) {
      yield put(requestVerificationEmail.failure(result));
    } else {
      yield put(requestVerificationEmail.success(result));
    }
  } catch (err) {
    yield put(requestVerificationEmail.failure(err));
  } finally {
    yield put(requestVerificationEmail.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(logout.TRIGGER, doLogout);
  yield takeLatest(requestVerificationEmail.TRIGGER, doRequestVerificationEmail);
}
