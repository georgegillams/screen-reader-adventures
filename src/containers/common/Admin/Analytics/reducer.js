import produce from 'immer';

import { load } from './actions';

export const initialState = {
  analytics: null,
  loading: false,
  loadError: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case load.REQUEST:
        draft.loading = true;
        draft.loadError = null;
        break;

      case load.SUCCESS:
        draft.loading = false;
        draft.analytics = payload;
        break;

      case load.FAILURE:
        draft.loading = false;
        draft.loadError = payload;
        break;
    }
  });

export default reducer;
