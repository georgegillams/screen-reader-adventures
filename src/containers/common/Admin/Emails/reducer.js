import produce from 'immer';

import { load, resend } from './actions';

export const initialState = {
  emails: null,
  loading: false,
  loadError: null,

  emailToResend: null,
  resending: false,
  resendError: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case load.REQUEST:
        draft.loading = true;
        draft.loadError = null;
        break;

      case load.SUCCESS:
        draft.loading = false;
        draft.emails = payload;
        break;

      case load.FAILURE:
        draft.loading = false;
        draft.loadError = payload;
        break;

      case resend.TRIGGER:
        draft.emailToResend = payload.emailToResend;
        break;

      case resend.REQUEST:
        draft.resending = true;
        draft.resendError = null;
        break;

      case resend.SUCCESS:
        draft.resending = false;
        break;

      case resend.FAILURE:
        draft.resending = false;
        draft.resendError = payload;
        break;
    }
  });

export default reducer;
