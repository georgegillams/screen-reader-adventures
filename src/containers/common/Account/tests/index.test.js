import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { initialState } from '../reducer';
import { initialState as initialAuthenticatorState } from 'containers/common/Authenticator/reducer';

import configureStore from 'client-utils/common/redux/configure-store';

import AccountIndex from '../index';
import Account from '../Container';

describe('<Account />', () => {
  let store;
  const spy = jest.fn();

  beforeAll(() => {
    store = configureStore({});
  });

  it('should render correctly - index', () => {
    const { container } = render(
      <Provider store={store}>
        <AccountIndex />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with initial state', () => {
    const { container } = render(
      <Provider store={store}>
        <Account
          logout={spy}
          requestVerificationEmail={spy}
          accountState={{
            ...initialState,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { name: 'userName' },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with initial state', () => {
    const { container } = render(
      <Provider store={store}>
        <Account
          logout={spy}
          requestVerificationEmail={spy}
          accountState={{
            ...initialState,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { name: 'userName' },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with logOutError', () => {
    const { container } = render(
      <Provider store={store}>
        <Account
          logout={spy}
          requestVerificationEmail={spy}
          accountState={{
            ...initialState,
            logOutError: { error: 'not_found', errorMessage: 'Invalid session' },
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { name: 'userName' },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with logOutResult', () => {
    const { container } = render(
      <Provider store={store}>
        <Account
          logout={spy}
          requestVerificationEmail={spy}
          accountState={{
            ...initialState,
            logOutResult: { success: 'logged out' },
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { name: 'userName' },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with loggingOut=true', () => {
    const { container } = render(
      <Provider store={store}>
        <Account
          logout={spy}
          requestVerificationEmail={spy}
          accountState={{
            ...initialState,
            loggingOut: true,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { name: 'userName' },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with requestingVerificationEmail=true', () => {
    const { container } = render(
      <Provider store={store}>
        <Account
          logout={spy}
          requestVerificationEmail={spy}
          accountState={{
            ...initialState,
            requestingVerificationEmail: true,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { name: 'userName' },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with requestVerificationEmailResult', () => {
    const { container } = render(
      <Provider store={store}>
        <Account
          logout={spy}
          requestVerificationEmail={spy}
          accountState={{
            ...initialState,
            requestVerificationEmailResult: { success: 'Verification email sent' },
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { name: 'userName' },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with requestVerificationEmailError', () => {
    const { container } = render(
      <Provider store={store}>
        <Account
          logout={spy}
          requestVerificationEmail={spy}
          accountState={{
            ...initialState,
            requestVerificationEmailError: { error: 'not_found', errorMessage: 'Invalid session' },
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { name: 'userName' },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with user with emailVerified=true', () => {
    const { container } = render(
      <Provider store={store}>
        <Account
          logout={spy}
          requestVerificationEmail={spy}
          accountState={{
            ...initialState,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { name: 'userName', emailVerified: true },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with no user', () => {
    const { container } = render(
      <Provider store={store}>
        <Account
          logout={spy}
          requestVerificationEmail={spy}
          accountState={{
            ...initialState,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
