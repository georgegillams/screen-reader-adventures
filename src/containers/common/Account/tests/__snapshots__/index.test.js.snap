// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Account /> should render correctly - index 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="skeleton__outer skeleton__section"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__button"
      style="margin-top: 1.2rem;"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__button"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
  </div>
</div>
`;

exports[`<Account /> should render correctly with initial state 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Account
    </h1>
    <div
      class="account__control-panel"
    >
      <button
        class="button__outer button__outer--regular button__outer--no-bouncy account__button"
        type="button"
      >
        <span
          class="button__child"
        >
          Get a new verification email
        </span>
      </button>
      <button
        class="button__outer button__outer--destructive button__outer--no-bouncy button__outer--no-bouncy--destructive"
        type="button"
      >
        <span
          class="button__child"
        >
          Logout
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<Account /> should render correctly with initial state 2`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Account
    </h1>
    <div
      class="account__control-panel"
    >
      <button
        class="button__outer button__outer--regular button__outer--no-bouncy account__button"
        type="button"
      >
        <span
          class="button__child"
        >
          Get a new verification email
        </span>
      </button>
      <button
        class="button__outer button__outer--destructive button__outer--no-bouncy button__outer--no-bouncy--destructive"
        type="button"
      >
        <span
          class="button__child"
        >
          Logout
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<Account /> should render correctly with logOutError 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Account
    </h1>
    <div
      class="account__control-panel"
    >
      <button
        class="button__outer button__outer--regular button__outer--no-bouncy account__button"
        type="button"
      >
        <span
          class="button__child"
        >
          Get a new verification email
        </span>
      </button>
      <button
        class="button__outer button__outer--destructive button__outer--no-bouncy button__outer--no-bouncy--destructive"
        type="button"
      >
        <span
          class="button__child"
        >
          Logout
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<Account /> should render correctly with logOutResult 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Account
    </h1>
    <div
      class="account__control-panel"
    >
      <button
        class="button__outer button__outer--regular button__outer--no-bouncy account__button"
        type="button"
      >
        <span
          class="button__child"
        >
          Get a new verification email
        </span>
      </button>
      <button
        class="button__outer button__outer--destructive button__outer--no-bouncy button__outer--no-bouncy--destructive"
        type="button"
      >
        <span
          class="button__child"
        >
          Logout
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<Account /> should render correctly with loggingOut=true 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Account
    </h1>
    <div
      class="account__control-panel"
    >
      <button
        class="button__outer button__outer--regular button__outer--no-bouncy account__button"
        type="button"
      >
        <span
          class="button__child"
        >
          Get a new verification email
        </span>
      </button>
      <button
        class="button__outer button__outer--destructive button__outer--no-bouncy button__outer--no-bouncy--destructive button__outer--disabled"
        disabled=""
        type="button"
      >
        <span
          class="button__child button__child--loading"
        >
          Logout
        </span>
        <div
          aria-hidden="true"
          class="button__loading-spinner"
        >
          <div
            class="spinner__outer"
          >
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
          </div>
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`<Account /> should render correctly with no user 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="skeleton__outer skeleton__section"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__button"
      style="margin-top: 1.2rem;"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
    <div
      class="skeleton__outer skeleton__button"
    >
      <div
        class="skeleton__shimmer"
        style="margin-left: -0px;"
      />
    </div>
  </div>
</div>
`;

exports[`<Account /> should render correctly with requestVerificationEmailError 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Account
    </h1>
    <div
      class="account__control-panel"
    >
      <button
        class="button__outer button__outer--regular button__outer--no-bouncy account__button"
        type="button"
      >
        <span
          class="button__child"
        >
          Get a new verification email
        </span>
      </button>
      <button
        class="button__outer button__outer--destructive button__outer--no-bouncy button__outer--no-bouncy--destructive"
        type="button"
      >
        <span
          class="button__child"
        >
          Logout
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<Account /> should render correctly with requestVerificationEmailResult 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Account
    </h1>
    <div
      class="account__control-panel"
    >
      <button
        class="button__outer button__outer--regular button__outer--no-bouncy account__button"
        type="button"
      >
        <span
          class="button__child"
        >
          Get a new verification email
        </span>
      </button>
      <button
        class="button__outer button__outer--destructive button__outer--no-bouncy button__outer--no-bouncy--destructive"
        type="button"
      >
        <span
          class="button__child"
        >
          Logout
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<Account /> should render correctly with requestingVerificationEmail=true 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Account
    </h1>
    <div
      class="account__control-panel"
    >
      <button
        class="button__outer button__outer--regular button__outer--no-bouncy button__outer--disabled account__button"
        disabled=""
        type="button"
      >
        <span
          class="button__child button__child--loading"
        >
          Get a new verification email
        </span>
        <div
          aria-hidden="true"
          class="button__loading-spinner"
        >
          <div
            class="spinner__outer"
          >
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
            <div
              class="spinner__inner"
            />
          </div>
        </div>
      </button>
      <button
        class="button__outer button__outer--destructive button__outer--no-bouncy button__outer--no-bouncy--destructive"
        type="button"
      >
        <span
          class="button__child"
        >
          Logout
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<Account /> should render correctly with user with emailVerified=true 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="page-title__heading"
    >
      Account
    </h1>
    <div
      class="account__control-panel"
    >
      <button
        class="button__outer button__outer--destructive button__outer--no-bouncy button__outer--no-bouncy--destructive"
        type="button"
      >
        <span
          class="button__child"
        >
          Logout
        </span>
      </button>
    </div>
  </div>
</div>
`;
